### DISCUSSION TITLE
Subscription Context

### DISCUSSION THREAD

**Discussion by lucapinelli** [enisdenjo/graphql-ws#207] [Q&A] [answered]
Hi,
I'm running a server implemented using the recipe: "ws server usage with Express GraphQL". During the `onConnect` I'm doing the user authentication. Is there a way to inject the user info (or any other info) in the GraphQL subscription context? If during the `onConnect` it is not possible, in which stage it is possible to modify the context that will be available during the GraphQL subscriptions?

**Comment by enisdenjo** [comment, marked as answer]
Yap, you can use [the `context` server option](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.serveroptions.md#context) (there is also a ["ws server usage with custom context value" recipe](https://github.com/enisdenjo/graphql-ws#context)). You can either provide it with a static context or a function which returns a context.

This value is then used as the GraphQL execution arguments `contextValue`:
https://github.com/enisdenjo/graphql-ws/blob/79bd4290514b75691acaf3bf2c6e5e710b787eb1/src/server.ts#L718-L723

**Reply by lucapinelli** [reply]
thank you! This works perfectly :+1:

**Reply by bnussman** [reply]
I'm so confused. I want to set the context `onConnect`. The accepted answer does not tell you how to set it **only** onConnect. I don't want my auth checking to happen for "each operation".

**Reply by enisdenjo** [reply]
@bnussman you cannot set the GraphQL context in `onConnect`, only in [`context`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#context) and [`onSubscribe`](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/server.ServerOptions.md#onsubscribe).

If you don't want the re-create the context for each operation; you can create one in `onConnect`, set it on the connection context extra and then just forward it down inside `context`. Roughly:
```js
import { authenticateUser } from './my-auth';

useServer(
  {
    // authenticate the user and set it on the connection context
    onConnect: async (ctx) => {
      const user = await authenticateUser(ctx);
      if (!user) {
        return false;
      }
      ctx.extra.user = user;
    },
    // pass down the user from the connection context down to the GraphQL context
    context: (ctx) => ({
      user: ctx.extra.user,
    }),
  },
);
```

**Reply by bnussman** [reply]
Thank you so much. I totally neglected to see the `extra` field. Appreciate you @enisdenjo!
